### ISSUE TITLE
queue_order_strategy = priority randomizes queue order

### ISSUE THREAD

**Issue by crabhi** [celery/celery#8673]
# Checklist

- [x] I have verified that the issue exists against the `main` branch of Celery.
- [ ] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first.
- [x] I have read the relevant section in the
  [contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `main` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- #5127 

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: main (5.3.6)

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```

software -> celery:5.3.6 (emerald-rush) kombu:5.3.4 py:3.10.12
            billiard:4.2.0 redis:5.0.1
platform -> system:Linux arch:64bit, ELF
            kernel version:6.2.0-37-generic imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled

broker_url: 'redis://127.0.0.1:6379/0'
deprecated_settings: None
broker_transport_options: 
 'queue_order_strategy': 'priority'}

```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

- **Minimal Python Version**: N/A or Unknown
- **Minimal Celery Version**: N/A or Unknown
- **Minimal Kombu Version**: N/A or Unknown
- **Minimal Broker Version**: N/A or Unknown
- **Minimal Result Backend Version**: N/A or Unknown
- **Minimal OS and/or Kernel Version**: N/A or Unknown
- **Minimal Broker Client Version**: N/A or Unknown
- **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==5.2.0
async-timeout==4.0.3
billiard==4.2.0
celery @ file:///home/krab/workspace/celery
click==8.1.7
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.3.0
kombu==5.3.4
prompt-toolkit==3.0.41
python-dateutil==2.8.2
redis==5.0.1
six==1.16.0
tzdata==2023.3
vine==5.1.0
wcwidth==0.2.12

```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

`proj.py`:

```python
import time

from celery import Celery

app = Celery('proj', broker='redis://127.0.0.1:6379/0')
app.conf.update(broker_transport_options={
    'queue_order_strategy': 'priority'
})

@app.task
def t(q):
    print(f'{q} start')
    time.sleep(1)

```

Tasks scheduling:

```python
In [1]: from proj import t

In [2]: for i in range(5000):
   ...:     q = 'high' if i %2 == 0 else 'low'
   ...:     t.apply_async([q], queue=q)
```

Worker CLI:

```
celery -A proj worker -Q high,low --concurrency 1
```

Start the worker, let it process a few tasks and kill it. Repeat and observe which tasks get processed.

</p>
</details>

# Expected Behavior
I expect that every time, tasks from the `high` queue are consumed first, then the tasks from the `low` queue.

# Actual Behavior

The worker always consumes tasks from one queue first. However, which one it is varies between restarts. If I start `celery -A proj worker -Q high,low --concurrency 1`, it may consume tasks from `high`. When I kill it and start again, it may start picking tasks from `low` instead.

I think the culprit may be that `kombu.transport.redis.Channel` uses a set for its `active_queues` property. Therefore the strategy itself receives the queues in effectively randomized order.

**Comment by yoppi** [comment]
@crabhi 
You can specify the order of the queues by specifying options.
see. https://github.com/celery/kombu/blob/2d6d8baf0f9cf6812c70b7ddb06b242bf31c8ad0/kombu/transport/redis.py#L673-L678

**Comment by crabhi** [comment]
Yes, that's the workaround we use. Still, the behaviour of `priority` is - I would say - broken.

**Comment by shkai** [comment]
Does rabbitmq have a similar solution?  -Q q1,q2     q2 can work when q1 is empty
